traitlets/traitlets.py

```diff
@@ -644,6 +648,17 @@ def hold(*a):
                 for v in cache.values():
                     self._notify_trait(*v)
 
+    @contextlib.contextmanager
+    def _allow_write(self):
```

Comment:
That is right. It would make sense to have a different setter for read-only attributes. It might still require a boolean check in `__set__` though.